### PR TITLE
add report to srsnv training

### PR DESCRIPTION
**Pull request by rinashalibo**

### CODE REVIEW

File: test/resources/unit/mrd/test_mrd_utils/synthetic_signature_test.vcf.gz.csi
Comment (kenissur) [inline test/resources/unit/mrd/test_mrd_utils/synthetic_signature_test.vcf.gz.csi:1]: Please generate a .tbi index for consistency (bcftools index -t)

File: test/resources/unit/mrd/test_srsnv_plotting_utils/balanced_ePCR_LA5_LA6_333_LuNgs_08.params.json
Comment (kenissur) [inline test/resources/unit/mrd/test_srsnv_plotting_utils/balanced_ePCR_LA5_LA6_333_LuNgs_08.params.json:1]: Please reformat the json file so it is readable

File: ugvc/mrd/srsnv_training_utils.py [hunk 1]
```diff
@@ -365,6 +368,10 @@ class from which classifier will be instantiated, by default xgb.XGBClassifier
         balanced_sampling_info_fields : list[str], optional
             List of info fields to balance the TP data by, default None (do not balance)
             Recommended in order to avoid the motif distribution of germline variants being learned (data leak)
+        lod_filters : dict, optional
+            dict of format 'filter name':'querie' for LoD simulation, by default None
```
Comment (kenissur) [inline ugvc/mrd/srsnv_training_utils.py]: query

File: ugvc/mrd/srsnv_training_utils.py [hunk 2]
```diff
@@ -325,6 +326,8 @@ def __init__(
         model_params: dict | str = None,
         classifier_class=xgb.XGBClassifier,
         balanced_sampling_info_fields: list[str] = None,
+        lod_filters: dict = None,
+        adapter_version: str = None,
```
Comment (kenissur) [inline ugvc/mrd/srsnv_training_utils.py]: please rename to balanced_strand_adapter_version

File: ugvc/pipelines/mrd/srsnv_training.py [hunk 3]
```diff
@@ -116,6 +117,20 @@ def parse_args(argv: list[str]) -> argparse.Namespace:
         default=DEFAULT_FLOW_ORDER,
         help="""flow order - required for cycle skip annotation """,
     )
+    parser.add_argument(
+        "--lod_filters",
```
Comment (kenissur) [inline ugvc/pipelines/mrd/srsnv_training.py:121]: This is a very long input so I think it's better to take a json file instead of an explicit dictionary

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 4]
```diff
@@ -0,0 +1,1541 @@
...
+    probs = classifier.predict_proba(X[cls_features])
+    predictions = classifier.predict(X[cls_features])
+    quals = -10 * np.log10(1 - probs)
+    predictions_df = pd.DataFrame(y)
+    labels = np.unique(y["label"].astype(int))
+    for label in labels:
+        predictions_df[f"ML_prob_{label}"] = probs[:, label]
+        predictions_df[f"ML_qual_{label}"] = quals[:, label]
+        predictions_df[f"ML_prediction_{label}"] = predictions[:, label]
+    df = pd.concat([X, predictions_df], axis=1)
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py:90]: TODO: use the information from adapter_version instead of this patch

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 5]
```diff
@@ -0,0 +1,1541 @@
...
+    "BQ60": f"{FeatureMapFields.X_SCORE.value} >= 6 and X_EDIST <= 5",
+    "BQ80": f"{FeatureMapFields.X_SCORE.value} >= 7.9 and X_EDIST <= 5",
+    "CSKP": f"{FeatureMapFields.X_SCORE.value} >= 10 and X_EDIST <= 5",
+    "BQ80_trim_ends": f"{FeatureMapFields.X_SCORE.value} >= 7.9 and X_EDIST <= 5 and X_INDEX > 12 \
+        and X_INDEX < (X_LENGTH - 12)",
+    "CSKP_trim_ends": f"{FeatureMapFields.X_SCORE.value} >= 10 and X_EDIST <= 5 and X_INDEX > 12 \
+        and X_INDEX < (X_LENGTH - 12)",
+    "BQ60_mixed_only": f"{FeatureMapFields.X_SCORE.value} >= 6 and X_EDIST <= 5 and is_mixed",
+    "BQ80_mixed_only": f"{FeatureMapFields.X_SCORE.value} >= 7.9 and X_EDIST <= 5 and is_mixed",
+    "BQ80_mixed_CSKP_only": f"{FeatureMapFields.X_SCORE.value} >= 7.9 and X_EDIST <= 5 and is_mixed and is_cycle_skip",
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py]: Please remove the BQ80_mixed_CSKP_only filter, BQ80 AND CSKP_only do not make sense together (CSKP is also BQ80)

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 6]
```diff
@@ -0,0 +1,1541 @@
...
+    recalls : dict
+        list of false positive rates per ML score per label
+    """
+
+    cls_features = classifier.feature_names_in_
+    probs = classifier.predict_proba(X[cls_features])
+    predictions = classifier.predict(X[cls_features])
+    quals = -10 * np.log10(1 - probs)
+    predictions_df = pd.DataFrame(y)
+    labels = np.unique(y["label"].astype(int))
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py:84]: TODO: write the code below with .assign rather than concat

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 7]
```diff
@@ -0,0 +1,1541 @@
...
+        plt.savefig(
+            output_filename,
+            facecolor="w",
+            dpi=300,
+            bbox_inches="tight",
+            bbox_extra_artists=[title_handle, legend_handle],
+        )
+
+
+def LoD_training_summary(
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py:248]: why "training"?

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 8]
```diff
@@ -0,0 +1,1541 @@
...
+from scipy.stats import binom
+from sklearn.metrics import confusion_matrix, precision_score, recall_score
+from tqdm import tqdm
+
+from ugvc import logger
+from ugvc.mrd.balanced_strand_utils import BalancedStrandAdapterVersions
+from ugvc.mrd.featuremap_utils import FeatureMapFields
+from ugvc.utils.metrics_utils import read_effective_coverage_from_sorter_json
+from ugvc.utils.misc_utils import exec_command_list
+
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py:23]: TODO: use constants instead of X_EDIST etc.

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 9]
```diff
@@ -0,0 +1,1541 @@
...
+        ),
+        axis=1,
+    )
+
+    df_summary = df_summary.assign(precision=df_summary["TP"] / (df_summary["FP"] + df_summary["TP"]))
+
+    # save outputs
+    params_for_save = {"residual_snv_rate_no_filter": residual_snv_rate_no_filter}
+    with open(output_params_file, "w", encoding="utf-8") as f:
+        json.dump(params_for_save, f)
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py:352]: TODO: it makes more sense to calculate the error rate column here than to carry it over to another function and save a single scalar value separately. I mean this code:
```
df_mrd_sim = (df_summary[["TP"]]).rename(columns={"TP": "read_retention_ratio"})
    df_mrd_sim = df_mrd_sim.assign(
        residual_snv_rate=df_summary["FP"] * residual_snv_rate_no_filter / df_mrd_sim["read_retention_ratio"]
    )
```

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 10]
```diff
@@ -0,0 +1,1541 @@
...
+def plot_LoD(
+    df_mrd_sim: pd.DataFrame,
+    lod_label: str,
+    c_lod: str,
+    filters: dict,
+    adapter_version: str,
+    min_LoD_filter: str,
+    title: str = "",
+    output_filename: str = None,
+    fs: int = 14,
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py]: please rename to font_size

File: ugvc/mrd/srsnv_plotting_utils.py [hunk 11]
```diff
@@ -0,0 +1,1541 @@
...
+        the filter which minimizes the LoD
+    title : str, optional
+        title for the generated plot, by default ""
+    output_filename : str, optional
+        path to which the plot will be saved, by default None
+    fs : int, optional
+        font size for the plot, by default 14
+
+    """
+
```
Comment (kenissur) [inline ugvc/mrd/srsnv_plotting_utils.py]: TODO: a less patchy solution, perhaps accept ml_filters as a separate input

File: ugvc/mrd/srsnv_training_utils.py [hunk 12]
```diff
@@ -483,11 +493,30 @@ def save_model_and_data(self):
             "train_set_size": self.train_set_size,
             "test_set_size": self.test_set_size,
             "fp_featuremap_entry_number": self.fp_featuremap_entry_number,
+            "lod_filters": self.lod_filters,
+            "adapter_version": self.adapter_version,
+            "columns": self.columns,
         }
 
         with open(self.params_save_path, "w", encoding="utf-8") as f:
             json.dump(params_to_save, f)
```
Comment (kenissur) [inline ugvc/mrd/srsnv_training_utils.py:504]: TODO: merge the two reports so train and test set results are presented together

Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py:504]: TODO: change the report so it produces metrics that are saved into a h5 table in addition to the plots

**Review by kenissur — APPROVED** [review]
Hi,

I gave only minor, easy to fix comments, please address them. Additionally, there are more major issues I would like to address in the future, I added comments starting with "TODO" - please copy these into "TODO" comments in the code itself.

Let's merge asap!

File: test/resources/unit/mrd/test_mrd_utils/synthetic_signature_test.vcf.gz.csi
Comment (rinashalibo) [reply test/resources/unit/mrd/test_mrd_utils/synthetic_signature_test.vcf.gz.csi:1]: tried and got this error:
(genomics.py3) rina.schwartz@dv-rinag /data1/proj/VariantCalling/test/resources/unit/mrd/test_mrd_utils (rinas/add_report_to_master) $ bcftools index -t synthetic_signature_test.vcf.gz
[E::bcf_index_build3] TBI indices for BCF files are not supported
index: failed to create index for "synthetic_signature_test.vcf.gz"

I'll ask Gat about this file, I think its hers

File: test/resources/unit/mrd/test_srsnv_plotting_utils/balanced_ePCR_LA5_LA6_333_LuNgs_08.params.json
Comment (rinashalibo) [reply test/resources/unit/mrd/test_srsnv_plotting_utils/balanced_ePCR_LA5_LA6_333_LuNgs_08.params.json:1]: done

File: ugvc/mrd/srsnv_training_utils.py (hunk 1 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_training_utils.py]: done

File: ugvc/mrd/srsnv_training_utils.py (hunk 2 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_training_utils.py]: done

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 4 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py:90]: added, 
but I think its safer to use a condition on the specific fields (keep it like it is now) to avoid a crash. WDYT?

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 5 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py]: done

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 7 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py:248]: because it generates a summary of the training results

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 11 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py]: added

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 9 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py:352]: added

File: ugvc/pipelines/mrd/srsnv_training.py (hunk 3 above)
Comment (rinashalibo) [reply ugvc/pipelines/mrd/srsnv_training.py:121]: done

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 8 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py:23]: added

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 10 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py]: done

File: ugvc/mrd/srsnv_plotting_utils.py (hunk 6 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_plotting_utils.py:84]: added

File: ugvc/mrd/srsnv_training_utils.py (hunk 12 above)
Comment (rinashalibo) [reply ugvc/mrd/srsnv_training_utils.py:504]: added
